### ISSUE TITLE
Village cleanup and improvements

### ISSUE THREAD

**Issue by topisani** [Team-RTG/1.8.9-Realistic-Terrain-Generation#30]
Im currently working on a few village improvements. The exact scope of this is to be determined, but in the beginning it will just be trying to improve the villagematerials code and minimizing burried houses and the like. Later i might also add bridges over water and the like.

- [x] Replace VillageStructurePieces with a custom version
- [x] Integrate villagematerials with that new VillageStructurePiecesRTG, fixing doors and wells, and maybe adding support for more block replacements.
- [ ] Add a system for writing block configuration on one line, using the syntax in the comment  below.
- [ ] Completely revamp villagematerials to use presets instead of using settings for all individual blocks for each biome. I.e you choose a customisable preset for each biome instead of actually setting each block

**Comment by topisani** [comment]
The good thing about the villagematerial integration, is that instead of having an option to replace all cobblestone, we can have a "church block" option for example. This makes sure things can make a little bit more sense, as currently the "Roof Block" option sets both the roof and the chairs inside, since they are the same block.

**Comment by topisani** [comment]
A bit more in depth on the preset thing.
This means in biomeconfig you would have an option called "Village Material" or something. You would then set that to either a default preset or your own, for example "ACACIA" or "EL_DORADO"
Then in `config/rtg/villages.conf` you have options for presets. a such preset could look like this:

```
         EL_DORADO {
            #  [default: ]
            S:"Blacksmith Roof Block (Stone Slabs)"=minecraft:gold_block/0

            #  [default: ]
            S:"Corner Block (Logs)"=minecraft:gold_block/0

            #  [default: ]
            S:"Door Block (Oak door)"=minecraft:gold_block/0

            #  [default: ]
            S:"Fence Block (Wooden Fencepost)"=minecraft:gold_block/0

            #  [default: ]
            S:"Foundation block (Cobblestone)"=minecraft:gold_block/0

            #  [default: ]
            S:"Path Block (Gravel)"=minecraft:gold_block/0

            #  [default: ]
            S:"Roof Block (Wooden Stairs)"=minecraft:gold_block/0

            #  [default: ]
            S:"Stairs Block (Cobblestone Stairs)"=minecraft:gold_block/0

            #  [default: ]
            S:"Wall Block (Planks)"=minecraft:gold_block/0
        }
```

**Comment by topisani** [comment]
now this is posible:
![2016-02-29_16 21 16](https://cloud.githubusercontent.com/assets/3133596/13398874/a6d8fe1a-df00-11e5-98b0-55bd4cbd53db.png)
That is because i am messing with the generation of the church itself. Opens up a whole new world of possibilities!

**Comment by Zeno410** [comment]
This is all great stuff, but it should be a separate mod from RTG.

**Comment by Razorskills** [comment]
"RTG Addon: Modular Vilages"

**Comment by topisani** [comment]
@Zeno410 well, im talking to pink about it, if its not clear the point is not doing something like the image above. its simply to make villages of biome specific materials, in a better way. The bridges i mentioned was just an idea i got from how its done in vanilla minecraft PE and maybe console versions too.
I definately agree this could easily be expanded into its own mod, but i actually think it fits pretty well into RTG. as long as its not about custom villages, but realistic villages. 
Also of cause i will not be messing with the actual formation or function of the structures, and the image above is not in any way intended to be default behavior ;)

**Comment by whichonespink44** [comment]
I'll admit... I'm torn.

Part of me thinks villages are within RTG's scope in the same way that surface decorations are within RTG's scope. After all, decorations aren't technically 'terrain' gen and yet they live happily within RTG's scope.

At the same time, part of me thinks that this should be an addon because villages just feel like they should be their own thing.

The main thing that's making me learn towards keeping them in core are the issues we have with desert villages. The solution we have at the moment is hit & miss. Most of the time it works, but depending on where the world-coord-check lands inside isDesertVillage() you can sometimes get villages that are half desert village, half normal village. And the well doesn't always work because of a Forge issue where the biome is always null for well. But I believe (and someone please correct me if I'm wrong) that implementing our own village system would allow us to definitively resolve those issues.

But I'm still on the fence, so I'd love to hear more pros and cons on this.

**Comment by topisani** [comment]
as i said i definately agree it could be its own mod. But i also dont think its misplaced. Yesterday we talked about what things fit in RTG, and one of the main points was to not alter the feel of the game, just the aesthetics of the world generation. i personally dont think this alters the feel, as long as its just replacing blocks biomewise. Also, the only way to fix the desert village thing is using at least some part of this system.

To sum up, i think the biome specific villages belong in RTG, and the config options are really just additional glitter, mod support and customisability.
Going further than changing the blocks to biome specific realistic ones probably doesnt belong here though, and i think that should be dropped.

**Comment by Zeno410** [comment]
The killer question is: should people who don't want an RTG world for any reason be denied a nice village system? Bluntly, no way, and not everybody wants the RTG worlds. There are a lot of people who like crazified terrain - even demand it. And what about villages in alternate dimensions?

The second issue is that by tying two mods together, each is restricted by the incompatibilities of the other. So if somebody likes a mod that can't run with RTG then they can't have the villages, and if somebody likes a mod that's not compatible with the villages they can't run RTG.

The only thing that I would consider appropriate is a fix for the split-village problem. That's really more of a Forge problem, as vanilla villages don't have it. The problem is that they don't pass all the information you need for a block replacement decisions - which should include the building and village the block's supposed to be a part of. Maybe there's something we can intercept the way we do the decoration event.

**Comment by topisani** [comment]
Like pink, im a bit torn on this.
I think theres no harm in having it in RTG, but i think your point @Zeno410 is very important. However, it would mean i had to put it into its own mod, making this quite a bigger workload for me. 
Also, i think its a pretty small feature for its own mod. Yes it would definately be basis for something bigger, but personally i dont have time or really the will to do that.
Im thinking, many of the people who want to use RTG would want to use this too, and if not, disabling it is just a config file away. I also think it fits, maybe on the edge, but i do think it fits inside RTG.
I might be able to do a seperate mod for this, and i might be able to keep it up to date along with RTG1.8.9. 
But its certainly a lot easier for me to keep it in there, and i dont think it feels misplaced

**Comment by Zeno410** [comment]
Making a separate mod is a couple hours work, tops, and then it's done. I don't find maintaining multiple mods particularly hard, and sometimes it's quite a relief to be able to fix one mod when you can't fix another. This is ESPECIALLY true for version updates.

As long as it's just making biome-specific modification to the palette it's a small mod but if you're splitting out different buildings it's not a small mod anymore, and it's prone to expand. Add in a feature for user-defined or even mod-added buildings a la Ruins and it could turn into a pretty big project.

**Comment by topisani** [comment]
closed as i ended up listening to Zeno. Prepare for the first RTG addon!

**Comment by whichonespink44** [comment]
more like, the first non-RTG mod by Team RTG ;)

**Comment by Zeno410** [comment]
I'm really looking forward to it. I love the concept of villages, but they're rather bland. The village mods I've played with -ancient warfare and Millenaire - go too far in other directions. But a mod which makes the churches look like the most special building in the village? Adds some variety to the village palettes? I'm all in. Expect a UBC addon too.
